### PR TITLE
Fix context issue

### PR DESCRIPTION
**Pull request by CharlyYu**
1. fix allocator address issue
    2. make Context in Task a arch agnostic struct
    3. use ttbr0_el1 as the init sp, no kernel sp used
    4. fix pagetable fork issue, add map of mmio address
    5. support context swap

### CODE REVIEW

File: qlib/kernel/arch/aarch64/arch_def.rs [hunk 1]
```diff
@@ -195,3 +195,67 @@ impl State {
...
+	pub x23: u64,
+	pub x24: u64,
+	pub x25: u64,
+	pub x26: u64,
+	pub x27: u64,
+	pub x28: u64,
+	pub fp: u64,
+	pub sp: u64,
+	pub pc: u64,
+    pub tls: u64,
```
Comment (shrik3) [inline qlib/kernel/arch/aarch64/arch_def.rs:217]: seems like a mix of space/tab indentation

**Comment by shrik3** [comment]
**clock arithmetic overflows** 

after some time of running (dozen seconds) this error message pops out:

```
[0/40004365] Unable to update params: Common("old now time computation overflowed2. params = Parameters { BaseCycles: 125214, BaseRef: 19777724432452382, Frequency: 1000000000 }, now = 1000108435").
```

The error is reported from 

```
qlib/kernel/kernel/timer/parameters.rs:149
```

**Comment by chl337** [comment]
Hi @CharlyYu, regarding FP initialisation, could you try this [patch](https://jmp.sh/s/v1oiIgdn4ahnKql19MB4) and let me know of any change.
Could you also post the result of `lscpu | grep "Flags"`

**Comment by CharlyYu** [comment]
Hi @chl337  this is the result on my raspberrypi
```
feng@raspberrypi:~$ lscpu | grep "Flags"
Flags:                              fp asimd evtstrm crc32 cpuid

```

**Comment by CharlyYu** [comment]
> could you try this [patch](https://jmp.sh/s/v1oiIgdn4ahnKql19MB4) and let me know of any change.
The patch works

**Comment by CharlyYu** [comment]
> clock arithmetic overflows

This must be a bug, and I will debug it later, but I think it is not related with this PR.

File: qlib/kernel/arch/aarch64/arch_def.rs [hunk 2]
```diff
@@ -195,3 +195,67 @@ impl State {
         }
     }
 }
+
+pub type Context = Aarch64Context;
+
+#[derive(Debug)]
+#[repr(C)]
+pub struct Aarch64Context {
```
Comment (SchwarzStein) [inline qlib/kernel/arch/aarch64/arch_def.rs:203]: is this the context to use for all situations ?

Comment (CharlyYu) [reply qlib/kernel/arch/aarch64/arch_def.rs:203]: I'm not quite sure, @chl337 is taking care of the context swap thing, he can add or delete registers from this struct if it is necessary.

Comment (chl337) [reply qlib/kernel/arch/aarch64/arch_def.rs:203]: As I understand it, this structure is used for task switching, e.g. after a scheduling decision... That said, AFAIK at least the _calle-saved, FP, LR, SP_ registers must be preserved, **ref:** [ABI](https://github.com/ARM-software/abi-aa/blob/2a70c42d62e9c3eb5887fa50b71257f20daca6f9/aapcs64/aapcs64.rst#general-purpose-registers), [Context Switch](https://developer.arm.com/documentation/den0024/a/The-Memory-Management-Unit/Context-switching).

**Comment by chl337** [comment]
Hi @CharlyYu, could you please check if this [patch](https://jmp.sh/s/V0N3LZExWyuA27sxo5vQ) works for you exactly the same as the status of this PR. It is meant to resolve possible future merge conflicts between us, as it seems we were touching some similar places. Please let me know in case of any error or preferences!
P.S. In case of errors, I can look at them first from Monday on. Sorry for the inconvenience!

File: qlib/kernel/arch/aarch64/arch_def.rs (hunk 2 above)
Comment (CharlyYu) [reply qlib/kernel/arch/aarch64/arch_def.rs:203]: Actually FP LR SP are included in the struct, only the LR is actually save in the field of name `pc`

Comment (SchwarzStein) [reply qlib/kernel/arch/aarch64/arch_def.rs:203]: wouldn't you want to save all the context x0-x31, especially when you do a conditional yield.
e.g.
```
 while( !check_asynchronous_event() ) {
     yield();
}
```

**Comment by SchwarzStein** [comment]
@CharlyYu , have you tested the refactored code in x86? Do we still run with no issues?

**Comment by CharlyYu** [comment]
> @CharlyYu , have you tested the refactored code in x86? Do we still run with no issues?

not yet

**Comment by SchwarzStein** [comment]
> > @CharlyYu , have you tested the refactored code in x86? Do we still run with no issues?
> 
> not yet

Any suggestions on how we proceed. 
Should the above be sufficient and I merge this or there is more you want to add to this?

File: qlib/kernel/arch/aarch64/arch_def.rs (hunk 2 above)
Comment (CharlyYu) [reply qlib/kernel/arch/aarch64/arch_def.rs:203]: I think not all the registers saved in X86 context too, I am not sure how the `yield` is supported, maybe it is supported by some signal mechanism.

**Comment by CharlyYu** [comment]
I am still working on the context swap thing, but currently it is blocked in the `LoadUserTable`. even I changed the codes to
```
#[inline]
pub fn LoadUserTable(table: u64) {
    unsafe {
        asm!("msr ttbr0_el1, {0}
        ic iallu
        dsb nsh
        isb
        ", 
        in(reg) table);
    };
}
```

The process is still blocked in this function call. I don't get the reason, maybe you guys can find it if it is convenient.

**Comment by CharlyYu** [comment]
> Should the above be sufficient and I merge this or there is more you want to add to this?

I have some code changes not submited. maybe we can merge this PR after the `LoadUserTable` thing get fixed.

**Comment by SchwarzStein** [comment]
> > Should the above be sufficient and I merge this or there is more you want to add to this?
> 
> I have some code changes not submited. maybe we can merge this PR after the `LoadUserTable` thing get fixed.

Could you make an rfc, and I will have a look at the entire changes for what might be causing this.
If you can point out the code you use to create the tables and configure the registers, I might spot the bug faster.

**Comment by CharlyYu** [comment]
@SchwarzStein I have got the reason, it is the forked pagetable did not map the mmio address.

**Comment by CharlyYu** [comment]
> Hi @CharlyYu, could you please check if this [patch](https://jmp.sh/s/V0N3LZExWyuA27sxo5vQ) works for you exactly the same as the status of this PR

@chl337 The context swap seems ok now with codes change in this PR now, I have implemented a preliminary version of `context_swap` assembly function in the `qlib/kernel/arch/aarch64/syscall_entey.s`. Could you please take a look at code changes in this PR? and there is any error or  omission in this imlemetation, it is welcome to leave you comment, or you can submit a new PR to correct them.

**Comment by CharlyYu** [comment]
> Should the above be sufficient and I merge this or there is more you want to add to this?

@SchwarzStein I think this PR is ready to be merged after clh337's approve. Please take a look, @chl337

**Comment by CharlyYu** [comment]
@chl337 Could you please take a special look at the memory barries instructions in the `context_swap`?  I am not sure if it is appropriate to do it like that.

**Pull request merged**